File: leetcode/2. problem/455.分发饼干.java
import java.util.Arrays;

/*
 * @lc app=leetcode.cn id=455 lang=java
 *
 * [455] 分发饼干
 */

// @lc code=start
class Solution {
    public int findContentChildren(int[] g, int[] s) {
        // 排序+贪心
        return extracted(g, s);
    }

    /**
     * 21/21 cases passed (7 ms)
     * Your runtime beats 100 % of java submissions
     * Your memory usage beats 60.18 % of java submissions (42.3 MB)
     * 
     * @param g
     * @param s
     * @return
     */
    private int extracted(int[] g, int[] s) {
        Arrays.sort(g);
        Arrays.sort(s);
        int cookies = 0;
        int childs = 0;
        while (cookies < s.length && childs < g.length) {
            if (s[cookies] >= g[childs]) {
                cookies++;
                childs++;
            } else {
                cookies++;
            }
        }
        return childs;
    }
}
// @lc code=end

File: leetcode/2. problem/golang/455.分发饼干.go
/*
 * @lc app=leetcode.cn id=455 lang=golang
 *
 * [455] 分发饼干
 */

// @lc code=start
func findContentChildren(g []int, s []int) (ans int) {
	return test(g, s)
}

func test(greed, size []int) (ans int) {
	sort.Ints(greed)
	sort.Ints(size)
	n, m := len(greed), len(size)
	for i, j := 0, 0; i < n && j < m; i++ {
		// 如果孩子的胃口大与饼干的size,
		for j < m && greed[i] > size[j] {
			j++
		}
		if j < m {
			ans++
			j++
		}
	}
	return
}
func method(g, s []int) (ans int) {
	sort.Ints(g)
	sort.Ints(s)
	n, m := len(g), len(s)
	for i, j := 0, 0; i < n && j < m; i++ {
		for j < m && g[i] > s[j] {
			j++
		}
		if j > m {
			ans++
			j++
		}
	}
	return
}

// @lc code=end

